src/OpenTelemetry/Internal/SelfDiagnosticsEventListener.cs

```diff
@@ -15,24 +15,142 @@
 
         public SelfDiagnosticsEventListener(EventLevel logLevel, SelfDiagnosticsConfigRefresher configRefresher)
         {
             this.logLevel = logLevel;
-            this.configRefresher = configRefresher;
+            this.configRefresher = configRefresher ?? throw new ArgumentNullException(nameof(configRefresher));
+        }
+
+        /// <summary>
+        /// Encode a string into the designated position in a buffer of bytes, which will be written as log.
+        /// If isParameter is true, wrap "{}" around the string.
+        /// The buffer should not be filled to full, leaving at least one byte empty space to fill a '\n' later.
+        /// If the buffer cannot hold all characters, truncate the string and replace extra content with "...".
+        /// The buffer is not guaranteed to be filled until the last byte due to variable encoding length of UTF-8,
+        /// in order to prioritize speed over space.
+        /// </summary>
+        /// <param name="str">The string to be encoded.</param>
+        /// <param name="isParameter">Whether the string is a parameter. If true, "{}" will be wrapped around the string.</param>
+        /// <param name="buffer">The byte array to contain the resulting sequence of bytes.</param>
+        /// <param name="position">The position at which to start writing the resulting sequence of bytes.</param>
+        /// <returns>The position of the buffer after the last byte of the resulting sequence.</returns>
+        internal static int EncodeInBuffer(string str, bool isParameter, byte[] buffer, int position)
+        {
+            int charCount = str.Length;
+            int ellipses = isParameter ? "{...}\n".Length : "...\n".Length;
+
+            // Ensure there is space for "{...}\n" or "...\n".
+            if (buffer.Length - position - ellipses < 0)
+            {
+                return position;
+            }
+
+            int estimateOfCharacters = (buffer.Length - position - ellipses) / 2;
+
+            // Ensure the UTF-16 encoded string can fit in buffer UTF-8 encoding.
+            // And leave space for "{...}\n" or "...\n".
+            if (charCount > estimateOfCharacters)
+            {
+                charCount = estimateOfCharacters;
+            }
+
+            if (isParameter)
+            {
+                buffer[position++] = (byte)'{';
+            }
+
+            position += Encoding.UTF8.GetBytes(str, 0, charCount, buffer, position);
+            if (charCount != str.Length)
+            {
+                buffer[position++] = (byte)'.';
+                buffer[position++] = (byte)'.';
+                buffer[position++] = (byte)'.';
+            }
+
+            if (isParameter)
+            {
+                buffer[position++] = (byte)'}';
+            }
+
+            return position;
+        }
+
+        internal void WriteEvent(string eventMessage, ReadOnlyCollection<object> payload)
```

Comment:
Would IReadOnlyCollection work? https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlycollection-1?view=netcore-3.1

Reply:
Yes, it would work. [`ReadOnlyCollection`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.readonlycollection-1?view=netcore-3.1) implements a bunch of interfaces including `IReadOnlyCollection`.

I used `ReadOnlyCollection` because it's the type of the [EventWrittenEventArgs.Payload Property](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventwritteneventargs.payload?view=netcore-3.1#System_Diagnostics_Tracing_EventWrittenEventArgs_Payload). Should I prioritize interfaces over base classes in definition of methods?